src/server/engine_shard.cc

```diff
@@ -735,10 +766,17 @@ void EngineShard::RetireExpiredAndEvict() {
     }
 
     // if our budget is below the limit
-    if (db_slice.memory_budget() < eviction_redline && GetFlag(FLAGS_enable_heartbeat_eviction)) {
+    if ((db_slice.memory_budget() < shard_memory_budget_threshold || rss_memory_to_free > 0) &&
+        GetFlag(FLAGS_enable_heartbeat_eviction)) {
       uint32_t starting_segment_id = rand() % pt->GetSegmentCount();
-      db_slice.FreeMemWithEvictionStep(i, starting_segment_id,
-                                       eviction_redline - db_slice.memory_budget());
+      size_t goal_bytes =
+          std::max(rss_memory_to_free, shard_memory_budget_threshold - db_slice.memory_budget());
```

Comment:
I see now that also the old logic for multiple dbs is broken because memory budget is update only when CacheStats is called (once in heartbeat)